### ISSUE TITLE
IIFE false positive when temporary lifetimes are involved

### ISSUE THREAD

**Issue by CAD97** [rust-lang/rust-clippy#8653]
### Summary

The `format_args!` macro creates temporaries dropped at the end of the expression, so

```rust
let x = format_args!("");
```

is almost always unusable. An IIFE can be used to extend the temporary lifetime to the entirety of the IIFE, allowing `format_args!` to be used for factoring out large `format_args!`.

This typically isn't necessary, as you an just nest `format_args!` directly, but is useful for cases where a `fmt;:Arguments` may be used in multiple places, e.g.

```rust
(|message| {
    #[cfg(debug_assertions)]
    panic!("{}", message);
    #[cfg(not(debug_assertions)]
    eprintln!("{}; attempting to continue...", message);
})(format_args!("big message that formats some locals and would be annoying to duplicate multiple times"));
```

### Lint Name

clippy::redundant_closure_call

### Reproducer

I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d28002c7cbf39cd170af8297c799f70f):

```rust
pub fn main() {
    let a = 2;
    let b = 3;
    let c = a + b;

    (|equasion| {
        println!("With Rust, we can know that {equasion}!");
    })(format_args!("{a} + {b} = {c}"));
}
```

I saw this happen:

```
warning: try not to call a closure in the expression where it is declared
 --> src/main.rs:6:5
  |
6 | /     (|equasion| {
7 | |         println!("With Rust, we can know that {equasion}!");
8 | |     })(format_args!("{a} + {b} = {c}"));
  | |_______________________________________^
  |
  = note: `#[warn(clippy::redundant_closure_call)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_call
```

I expected to see this happen:

No warnings (as with IIFE-for-`try`) because replacing the IIFE with a `let` binding changes temporary lifetimes to make the code no longer compile.

### Version

```text
rustc 1.61.0-nightly (76d770ac2 2022-04-02)
binary: rustc
commit-hash: 76d770ac21d9521db6a92a48c7b3d5b2cc535941
commit-date: 2022-04-02
host: x86_64-pc-windows-msvc
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by CAD97** [comment]
It's still probably not good code, but this case at least is more nuanced than "avoid IIFE".

<details><summary>The horrible code that I wrote to run into this originally</summary>

```rust
(|new|
    (|message| {
        cfg_if! {
            if #[cfg(feature = "debug_assertions")] {
                panic!("w{message}");
            } else if #[cfg(feature = "tracing")] {
                tracing::error!(parent: crate::span(), "W{message}");
            } else {
                eprintln!("W{message}");
            }
        }
    })
    (format_args!("hile handling the previous error, errored releasing {new}: {error}"))
)(format_args!("{}({new:p})", std::any::type_name::<R>()));
```
</details>

**Comment by kpreid** [comment]
Instead of an IIFE, you could use `match`:

```rust
match format_args!(...) {
    message => {
        eprintln!("{message}");
    }
}
```

This is the standard pattern when writing macros that need to extend the life of temporaries that might exist in an input expression, and I suspect it would generally be considered more idiomatic Rust than an IIFE. (And it won't provoke a lint.)

That said, it would be nice if Clippy made a recommendation of `let` or `match` based on whether or not the expression has (or might have) such temporaries, since this use of `match` is not at all obvious.
